### PR TITLE
Add check for TraceEnum_ELBO requirements

### PR DESCRIPTION
**Pull request by fritzo**
This adds a conservative warning to `TraceEnum_ELBO` in cases where the model or guide might contain dependencies on enumerated variables. The warning is only triggered in cases where the model or guide could easily be rewritten to be obviously correct.

### CODE REVIEW

File: pyro/util.py [hunk 1]
```diff
@@ -386,6 +387,56 @@ def check_site_shape(site, max_iarange_nesting):
...
+    enumerated_sites = set(name for name, site in guide_trace.nodes.items()
+                           if site["type"] == "sample" and site["infer"].get("enumerate"))
+    for role, trace in [('model', model_trace), ('guide', guide_trace)]:
+        irange_counters = {}
+        enumerated_contexts = defaultdict(set)
+        for name, site in trace.nodes.items():
+            if site["type"] != "sample":
+                continue
+            irange_counter = {f.name: f.counter for f in site["cond_indep_stack"] if not f.vectorized}
+            context = frozenset(f for f in site["cond_indep_stack"] if f.vectorized)
```
Comment (fritzo) [inline pyro/util.py]: We only care about vectorized frames, since only vectorized variables are enumerated in lock-step.

File: pyro/util.py [hunk 2]
```diff
@@ -386,6 +387,56 @@ def check_site_shape(site, max_iarange_nesting):
...
+
+def check_traceenum_requirements(model_trace, guide_trace):
+    """
+    Warn if user could easily rewrite the model or guide in a way that would
+    clearly avoid invalid dependencies on enumerated variables.
+
+    :class:`~pyro.infer.traceenum_elbo.TraceEnum_ELBO` enumerates over
+    synchronized products rather than full cartesian products. Therefore models
+    must ensure that no variable outside of an iarange depends on an enumerated
+    variable inside that iarange. Since full dependency checking is impossible,
```
Comment (martinjankowiak) [inline pyro/util.py]: ... is impossible without ____ ....

Comment (fritzo) [reply pyro/util.py]: full dependency checking is impossible without *an oracle to the halting problem*

**Comment by fritzo** [comment]
Any additional comments, or is this ready to merge?

**Comment by martinjankowiak** [comment]
yup, just waiting on a commitment from you to explain to any confused users why their model generated this warning : )

**Comment by fritzo** [comment]
yes, i hereby commit to responding to users about why warnings were generated.

**Pull request merged**